presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1716,6 +1718,15 @@ private boolean isJoinOnConditionReferencesRelatedFields(Expression expression,
                 return isJoinOnConditionReferencesRelatedFields(comparisonExpression.getLeft(), scope)
                         != isJoinOnConditionReferencesRelatedFields(comparisonExpression.getRight(), scope);
             }
+            if (expression instanceof ArithmeticBinaryExpression) {
+                ArithmeticBinaryExpression arithmeticBinaryExpression = (ArithmeticBinaryExpression) expression;
```

Comment:
Generalize this - write a function extractVariables and check that each side of the equality contains exactly one side.

Reply:
I rewrote this method completely.